### PR TITLE
TCN-618 automate `updateversion`

### PR DESCRIPTION
**Pull request by elliotmjackson**
Automating the first part of the buf release process:

#### Summary

1. Generate token
1. Checkout repository code
1. Install Buf
1. Update Buf Version
1. Create PR
1. Slack Notification

- Start by announcing that a release is being made in the Slack #general channel. During the release process, we should not be merging new commits to the `main` branch of the repositories listed above.

- Enter the `buf` repository and make sure you’re on the `HEAD` of `main`.

    ```bash
    cd "${BUFBUILD_BUF_DIR}"
    git switch main
    git pull origin main
    make installbuf
    buf --version
    ```

- Set the `${VERSION}` environment variable.

    ```bash
    # Replace this with the actual version, which should be the current version without the `-dev` suffix.
    # If this is a release candidate, add the `-rc` suffix and increment the release candidate.
    VERSION=1.1.0
    echo ${VERSION}
    ```

- The following steps will update the version in `private/buf/bufcli/bufcli.go` and then push the public version and tag:

    ```bash
    make updateversion VERSION=${VERSION}
    git status
    git diff
    
    git checkout -b release/v${VERSION}
    git commit -am "Update to v${VERSION}"
    git push origin release/v${VERSION}
    
    # Get that PR Approved and merged
    ```

#### Summary

1. Checkout repository code
1. Check changelog is modified

- Alter the “Unreleased” entry in `CHANGELOG.md` to point to `${VERSION}` and update the link at the bottom of the file. Use the format `YYYY-MM-DD` for the year.

    ```diff
    -## [Unreleased]
    +## [v1.1.0] - 2021-10-01
    ```

    ```diff
    -[Unreleased]: https://github.com/bufbuild/buf/compare/v1.1.0...HEAD
    +[v1.1.0]: https://github.com/bufbuild/buf/compare/v1.0.0...v1.1.0
    ```

this is a manual step, have included a github action to verify changelog modification

#### Summary

1. Generate token
1. Set VERSION variable from tag
1. Checkout repository code
1. Tag Release

Once the changelog has been modified and merged, a tag will be pushed - triggering the release event

```
git tag -a "v${VERSION}" -m "v${VERSION}"
git push origin "v${VERSION}"
```

#### Summary

1. Generate token
1. Checkout repository code
1. Set VERSION variable from tag
1. Create assets
1. unset keys
1. Set FILES variable from tag
1. Release
1. Slack Notification

```sh
make bufrelease
```

- At this point, you will have release assets at `"${GITHUB_WORKSPACE}/.build/release/buf/assets"` and will be uploaded to the release.

NOTE: the release content is currently auto generated, grabbing a block from the changelog will be possible automatically, just not immediately accessible to me.

#### Summary

1. Generate token
1. Checkout repository code
1. Set NEXT VERSION variable from tag
1. update version
1. update home brew badge
1. create PR

raises a PR to return the HEAD to dev state.

```sh
make updateversion VERSION=${{env.NEXT_VERSION}}

make updatehomebrewbadge
```

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.version != '' }}
+    steps:
+      - name: Generate token
+        id: generate_token
+        uses: tibdex/github-app-token@v1
+        with:
+          app_id: 251311
```
Comment (joshcarp) [inline .github/workflows/release.yaml:30]: what is app_id?

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          private_key: ${{ secrets.TOKEN_EXCHANGE_GH_APP_PRIVATE_KEY }}
```
Comment (joshcarp) [inline .github/workflows/release.yaml:31]: Is there any reason why there isn't some `ORG_GH_TOKEN` or something. maybe that's outdated and I'm just used to it

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      - uses: actions/setup-go@v3
+        with:
+          go-version: '^1.19.x'
+      - name: Install Buf
+        run: make installbuf
+      - name: Update Buf Version
+        run: make updateversion
+      - name: Create PR
+        id: cpr
+        uses: peter-evans/create-pull-request@ad43dccb4d726ca8514126628bec209b8354b6dd
```
Comment (joshcarp) [inline .github/workflows/release.yaml:48]: FYI all github runners have `git` and `gh` updated so the following also works:
```
git add .
git commit -m "message"
gh pr create --title "title" --fill
```

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+          title: "Release v${{env.VERSION}}"
+          body: |
+            Release prepared for ${{env.VERSION}}
+            
+            Reminder: Update the changelog
+          token: ${{ steps.generate_token.outputs.token }}
+      - name: Slack Notification
+        run: |
+          jq --null-input '{ text: "BufCLI Release v${{env.VERSION}} has started: ${{ steps.cpr.outputs.pull-request-url }}" }' \
+          | curl -sSL -X POST -H 'Content-Type: application/json' -d @- '${{ secrets.SLACK_RELEASE_NOTIFICATION_WEBHOOK }}'
```
Comment (joshcarp) [inline .github/workflows/release.yaml:63]: nice

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      - name: Tag Release
+        run: |
+          git config --global user.password ${{ steps.generate_token.outputs.token }}
+          git tag -d ${{env.VERSION}} 2> /dev/null || echo 'local ref does not exist'
+          git push origin :${{env.VERSION}} 2> /dev/null || echo 'remote ref does not exist'
+          git tag ${{env.VERSION}}
+          git push origin ${{env.VERSION}}
+  perform:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'push' && startsWith(github.ref_name, 'v') }}
```
Comment (joshcarp) [inline .github/workflows/release.yaml:114]:
```suggestion
    if: ${{ github.event_name == 'push' && startsWith(github.ref_name, 'v') && github.ref_type == 'tag'}}
```

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    steps:
+      - name: Generate token
+        id: generate_token
+        uses: tibdex/github-app-token@v1
+        with:
+          app_id: 251311
+          private_key: ${{ secrets.TOKEN_EXCHANGE_GH_APP_PRIVATE_KEY }}
+          repository: ${{ github.repository }}
+          permissions: >-
+            {"contents": "write", "pull_requests": "write"}
```
Comment (joshcarp) [inline .github/workflows/release.yaml:124]: being pretty pedantic but you could extract this into another job and put 

```
needs: generate-token
```
here

**Review by joshcarp — COMMENTED** [review]
mostly nit picks. Overall I don't think you could simplify it much more unless there's a github action that already does this (maybe you could make one?)

File: .github/workflows/release.yaml (hunk 2 above)
Comment (ElliotMJackson) [reply .github/workflows/release.yaml:31]: an "org token" is like a god token - this uses a github app to create a single use scoped token

File: .github/workflows/release.yaml (hunk 1 above)
Comment (ElliotMJackson) [reply .github/workflows/release.yaml:30]: uid of the github app installed on the repo

File: .github/workflows/release.yaml (hunk 6 above)
Comment (ElliotMJackson) [reply .github/workflows/release.yaml:124]: if i do this, i would need to pipe the resulting key into the `$GITHUB_OUTPUT` which would leak the secret.

**Comment by ElliotMJackson** [comment]
Resolves TCN-668

**Review by joshcarp — APPROVED** [review]
LGTM 🎉

**Pull request merged**